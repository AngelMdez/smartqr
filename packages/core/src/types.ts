export type SmartQRAction =
  | { type: 'deeplink'; url: string }
  | { type: 'web'; url: string }
  | { type: 'noop' };
