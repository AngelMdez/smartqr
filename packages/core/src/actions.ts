import { SmartQRAction } from './types';

export function decideAction(opts: { deeplink?: string; web?: string }): SmartQRAction | null {
  const { deeplink, web } = opts
  if (deeplink && deeplink.trim()) return { type: 'deeplink', url: deeplink }
  if (web && web.trim()) return { type: 'web', url: web }
  return null
}
