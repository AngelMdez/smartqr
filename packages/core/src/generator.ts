import QRCode from 'qrcode'
import { z } from 'zod'

export const GenerateQROptionsSchema = z.object({
  size: z.number().int().min(16).max(4096).default(128),
  margin: z.number().int().min(0).max(8).default(2),
  level: z.enum(['L', 'M', 'Q', 'H']).default('M'),
  darkColor: z.string().default('#000000'),
  lightColor: z.string().default('#FFFFFF'),
  version: z.number().int().min(1).max(40).optional(),
  transparentLight: z.boolean().default(false),
})

export type GenerateQROptions = z.input<typeof GenerateQROptionsSchema>

// Replace all occurrences of a given fill color with "none"
function makeLightTransparent(svg: string, lightHex: string) {
  // Normalize hex case for comparison in the SVG output
  const hex = lightHex.toLowerCase()
  // Replace fill="#hex" (hex may appear upper/lower in output)
  const rx = new RegExp(`fill=["']#?${hex.replace('#', '')}["']`, 'gi')
  return svg.replace(rx, 'fill="none"')
}

/**
 * Generate a QR Code as an SVG string.
 */
export async function generateQRCode(
  value: string,
  options?: GenerateQROptions
): Promise<string> {
  if (!value || typeof value !== 'string') {
    throw new Error('`value` must be a non-empty string')
  }

  const opts = GenerateQROptionsSchema.parse(options ?? {})
  // Always pass a valid hex to the renderer; make transparent afterwards if needed
  const lightHex = opts.lightColor

  const svgRaw = await QRCode.toString(value, {
    type: 'svg',
    errorCorrectionLevel: opts.level,
    margin: opts.margin,
    width: opts.size,
    version: opts.version,
    color: { dark: opts.darkColor, light: lightHex },
  })

  const svg = opts.transparentLight ? makeLightTransparent(svgRaw, lightHex) : svgRaw
  return svg.trim()
}
