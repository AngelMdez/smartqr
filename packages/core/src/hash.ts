// FNV-1a 32-bit (deterministic bucket for rollout)
export function hashFNV1a(input: string): number {
  let h = 0x811c9dc5
  for (let i = 0; i < input.length; i++) {
    h ^= input.charCodeAt(i)
    h = Math.imul(h, 0x01000193) >>> 0
  }
  return h >>> 0
}
