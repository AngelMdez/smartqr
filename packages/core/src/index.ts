export * from './rules'
export * from './evaluate'
export * from './resolve'
