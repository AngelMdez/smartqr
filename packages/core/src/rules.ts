import { z } from 'zod'

export const RulesSchema = z.object({
  version: z.number().default(1),
  default: z.object({
    web: z.string().url().optional(),
    ios: z.string().optional(),
    android: z.string().optional(),
    fallback: z.string().optional()
  }),
  routes: z.array(z.object({
    when: z.object({
      os: z.array(z.enum(['iOS','Android','Desktop'])).optional(),
      lang: z.array(z.string()).optional(),
      dateRange: z.tuple([z.string(), z.string()]).optional(), // YYYY-MM-DD
      rollout: z.object({ percentage: z.number().min(0).max(100), seed: z.string() }).optional()
    }).partial(),
    web: z.string().optional(),
    ios: z.string().optional(),
    android: z.string().optional(),
    fallback: z.string().optional()
  })).default([])
})

export type Rules = z.infer<typeof RulesSchema>
export type Context = { userId?: string; now?: number; lang?: string }
export type Target = { web?: string; ios?: string; android?: string; fallback?: string }
export type Evaluation = {
  os: 'iOS'|'Android'|'Desktop'
  lang: string
  target: Target
  reason: 'rule'|'default'
}
