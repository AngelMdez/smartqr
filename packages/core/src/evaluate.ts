import UAParser from 'ua-parser-js'
import { hashFNV1a } from './hash'
import type { Rules, Context, Evaluation } from './rules'

export function evaluateRules(rules: Rules, ctx: Context = {}): Evaluation {
  const ua = new UAParser().getResult()
  const os: 'iOS'|'Android'|'Desktop' =
    ua.os.name === 'iOS' ? 'iOS' : ua.os.name === 'Android' ? 'Android' : 'Desktop'

  const lang = (ctx.lang ?? (typeof navigator !== 'undefined' ? navigator.language : 'en')).slice(0, 2)
  const nowISO = new Date(ctx.now ?? Date.now()).toISOString().slice(0, 10) // YYYY-MM-DD

  const match = rules.routes.find(r => {
    const w = r.when ?? {}
    if (w.os && !w.os.includes(os)) return false
    if (w.lang && !w.lang.includes(lang)) return false
    if (w.dateRange) {
      const [from, to] = w.dateRange
      if (nowISO < from || nowISO > to) return false
    }
    if (w.rollout) {
      const { percentage, seed } = w.rollout
      const id = ctx.userId ?? 'anon'
      const bucket = hashFNV1a(id + seed) % 100
      if (bucket >= percentage) return false
    }
    return true
  })

  return { os, lang, target: (match ?? rules.default), reason: match ? 'rule' : 'default' }
}
