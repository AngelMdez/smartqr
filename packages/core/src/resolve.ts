// The resolver is responsible for taking a rules payload (JSON),
// evaluating it for the current context (OS/lang/date/rollout),
// and then attempting to execute the most appropriate action
// (deeplink on mobile; web on desktop; fallback after a timeout).

import { RulesSchema, type Rules, type Context, type Evaluation } from './rules'
import { evaluateRules } from './evaluate'

// We support two navigation "modes". `assign` keeps history,
// `replace` swaps the current entry (useful for resolvers).
type NavKind = 'assign' | 'replace'

// Public result returned to the caller (handy for logging/tests).
export type ResolveResult = {
  evaluation: Evaluation // what evaluateRules decided (os/lang/target/reason)
  deeplink?: string      // app scheme (myapp://...) chosen for the device
  web?: string           // web URL selected
  fallback?: string      // ultimate fallback if deeplink fails
  used: 'deeplink' | 'web' | 'fallback' | 'none' // what we actually tried in the end
  error?: unknown        // transport error, if any
}

// Public options to customize behavior + testability knobs.
export type ResolveOptions = {
  // How to fetch/obtain the rules JSON for a given id (e.g. "promo123").
  loadRules: (id: string) => Promise<unknown>
  // Optional explicit id; if omitted, we read it from `location.search`.
  id?: string
  // Extra evaluation context (userId/now/lang) to augment auto-detected values.
  context?: Context
  // Milliseconds to wait before considering the deeplink failed and firing fallback.
  // Typical values: 800–1500ms. Default: 1200ms.
  timeoutMs?: number
  // Desktop behavior: if true, go straight to web when OS === Desktop. Default: true.
  preferWebOnDesktop?: boolean
  // assign vs replace navigation strategy. Default: 'assign'.
  navigation?: NavKind
  // Lifecycle hooks around navigation. Fail-safe (exceptions ignored).
  onBefore?: (info: ResolveResult) => void
  onAfter?: (info: ResolveResult) => void
  // Injectable navigation fn (great for unit tests / headless envs).
  // Defaults to window.location.assign/replace.
  navigate?: (url: string, kind: NavKind) => void
}

// Extract `id` from the current URL (?id=...), with a safe fallback.
// Works in browsers; in SSR/Node we return "default".
function getIdFromLocation(): string {
  try {
    const href =
      typeof location !== 'undefined' ? location.href : 'http://localhost/?id=default'
    const id = new URL(href).searchParams.get('id')
    return id ?? 'default'
  } catch {
    // Malformed URL or no window/location available.
    return 'default'
  }
}

// Given an evaluation, pick the relevant URIs for this device.
// `evaluateRules` already normalized `target`, so we only branch on OS.
function chooseUris(e: Evaluation) {
  const { os, target } = e
  const deeplink =
    os === 'iOS' ? target.ios : os === 'Android' ? target.android : undefined
  const web = target.web
  const fallback = target.fallback ?? target.web // default fallback to web
  return { deeplink, web, fallback }
}

// Default navigation that hits the browser History API.
// Kept small and isolated so we can override it in tests.
function defaultNavigate(url: string, kind: NavKind) {
  if (typeof window === 'undefined' || typeof location === 'undefined') return
  if (kind === 'replace') window.location.replace(url)
  else window.location.assign(url)
}

/**
 * Load rules → evaluate → attempt to execute the best action.
 * This function is resilient to SSR, uses a timer heuristic for deeplink
 * fallback, and exposes hooks for analytics and custom navigation.
 */
export async function resolveAndExecute(opts: ResolveOptions): Promise<ResolveResult> {
  // 1) Normalize options and set defaults.
  const {
    id = getIdFromLocation(),
    loadRules,
    context,
    timeoutMs = 1200,
    preferWebOnDesktop = true,
    navigation = 'assign',
    onBefore,
    onAfter,
    navigate = defaultNavigate
  } = opts

  // 2) Load and validate the rules payload.
  //    Zod validation keeps the surface area tight and predictable.
  const raw = await loadRules(id)
  const rules = RulesSchema.parse(raw) as Rules

  // 3) Evaluate conditions (OS/lang/date/rollout) to pick a target.
  const evaluation = evaluateRules(rules, context)
  const { deeplink, web, fallback } = chooseUris(evaluation)

  // 4) Notify observers *before* any navigation occurs.
  const baseResult: ResolveResult = {
    evaluation,
    deeplink,
    web,
    fallback,
    used: 'none'
  }
  try {
    onBefore?.(baseResult)
  } catch {
    // Hooks must never break navigation; ignore exceptions on purpose.
  }

  // 5) Environment guard:
  // In non-browser environments we cannot use defaultNavigate (it needs window/location),
  // but if the caller injected a custom `navigate`, we can still proceed (e.g. tests, SSR).
  const isBrowser = typeof window !== 'undefined'
  const usingDefaultNav = navigate === defaultNavigate
  if (!isBrowser && usingDefaultNav) {
    const r = { ...baseResult, used: 'none' as const }
    try { onAfter?.(r) } catch {}
    return r
  }

  // 6) Desktop policy: if configured, prefer going directly to web.
  if (evaluation.os === 'Desktop' && preferWebOnDesktop && web) {
    navigate(web, navigation)
    const r = { ...baseResult, used: 'web' as const }
    try { onAfter?.(r) } catch {}
    return r
  }

  // 7) Mobile-first flow:
  //    Try the deeplink. If the app doesn't take focus quickly,
  //    assume it failed and go to `fallback` after `timeoutMs`.
  if (deeplink) {
    const isDoc = typeof document !== 'undefined'

    return await new Promise<ResolveResult>((resolve) => {
      let finished = false
      const settle = (used: ResolveResult['used']) => {
        if (finished) return
        finished = true
        const r: ResolveResult = { ...baseResult, used }
        try { onAfter?.(r) } catch {}
        resolve(r) // <-- resolve the promise right here
      }

      // Fallback timer
      const t = setTimeout(() => {
        if (!finished && fallback) {
          navigate(fallback, navigation)
          settle('fallback') // <-- resolve immediately on fallback
        }
      }, timeoutMs)

      try {
        // Try deeplink
        navigate(deeplink, navigation)

        // Cancel fallback if page loses visibility (app opened)
        const onVis = () => {
          if (isDoc && document.hidden) {
            clearTimeout(t)
            if (isDoc) document.removeEventListener('visibilitychange', onVis)
            settle('deeplink') // <-- resolve immediately on deeplink success
          }
        }
        if (isDoc) document.addEventListener('visibilitychange', onVis, { once: true })
      } catch {
        // navigate() threw -> go to fallback or finish
        clearTimeout(t)
        if (fallback) {
          navigate(fallback, navigation)
          settle('fallback')
        } else {
          settle('none')
        }
      }

      // Backstop: some WebViews never fire visibilitychange nor error
      setTimeout(() => {
        settle('deeplink') // best-effort assumption
      }, timeoutMs + 250)
    })
  }

  // 8) No deeplink path available → try web first, else fallback.
  if (web) {
    navigate(web, navigation)
    const r = { ...baseResult, used: 'web' as const }
    try { onAfter?.(r) } catch {}
    return r
  }
  if (fallback) {
    navigate(fallback, navigation)
    const r = { ...baseResult, used: 'fallback' as const }
    try { onAfter?.(r) } catch {}
    return r
  }

  // 9) Nothing to do (misconfigured rules). We still return a structured result.
  const r = { ...baseResult, used: 'none' as const }
  try { onAfter?.(r) } catch {}
  return r
}
