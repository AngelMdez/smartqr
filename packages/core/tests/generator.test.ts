import { describe, it, expect } from 'vitest'
import { generateQRCode } from '../src/generator'

describe('generateQRCode (SVG)', () => {
  it('returns a valid SVG string with defaults', async () => {
    const svg = await generateQRCode('hello-world-123')
    expect(typeof svg).toBe('string')
    expect(svg.startsWith('<svg')).toBe(true)
    expect(svg.includes('#000000')).toBe(true) // default dark color
  })

  it('respects size and colors', async () => {
    const svg = await generateQRCode('size-and-colors', {
      size: 256,
      margin: 1,
      darkColor: '#111111',
      lightColor: '#F0F0F0',
    })
    expect(svg.includes('#111111')).toBe(true)
    expect(svg.includes('#F0F0F0')).toBe(true)
    expect(svg.includes('256')).toBe(true)
  })

  it('can produce transparent light modules', async () => {
    const svg = await generateQRCode('transparent-light', {
      lightColor: '#EFEFEF',      // pass a valid hex; we will replace it to "none"
      transparentLight: true,
    })
    // Should not contain the light hex anymore
    expect(svg.toLowerCase().includes('#efefef')).toBe(false)
    // Should contain fill="none" for the light background/areas
    expect(svg.toLowerCase().includes('fill="none"')).toBe(true)
  })

  it('snapshot is stable with a fixed version', async () => {
    const svg = await generateQRCode('snapshot-stability', {
      version: 4,
      margin: 1,
      level: 'M',
    })
    expect(svg).toMatchSnapshot()
  })
})
