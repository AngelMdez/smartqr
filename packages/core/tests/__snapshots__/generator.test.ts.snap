// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateQRCode (SVG) > snapshot is stable with a fixed version 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="128" height="128" viewBox="0 0 35 35" shape-rendering="crispEdges"><path fill="#FFFFFF" d="M0 0h35v35H0z"/><path stroke="#000000" d="M1 1.5h7m1 0h2m4 0h1m1 0h4m1 0h1m1 0h1m2 0h7M1 2.5h1m5 0h1m5 0h1m2 0h1m1 0h1m2 0h3m1 0h1m1 0h1m5 0h1M1 3.5h1m1 0h3m1 0h1m2 0h3m1 0h1m5 0h2m1 0h1m3 0h1m1 0h3m1 0h1M1 4.5h1m1 0h3m1 0h1m1 0h2m2 0h1m1 0h1m2 0h5m1 0h2m1 0h1m1 0h3m1 0h1M1 5.5h1m1 0h3m1 0h1m1 0h2m1 0h1m1 0h2m3 0h4m2 0h1m1 0h1m1 0h3m1 0h1M1 6.5h1m5 0h1m1 0h1m1 0h2m1 0h2m1 0h1m1 0h2m6 0h1m5 0h1M1 7.5h7m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h7M9 8.5h1m5 0h5m2 0h1m1 0h1M1 9.5h1m3 0h1m1 0h6m1 0h2m1 0h1m1 0h2m1 0h2m1 0h6m2 0h1M1 10.5h1m1 0h1m2 0h1m2 0h1m1 0h4m5 0h2m2 0h1m3 0h1m1 0h1m1 0h2M4 11.5h2m1 0h1m2 0h4m2 0h1m1 0h1m2 0h7m2 0h4M3 12.5h2m3 0h2m2 0h1m2 0h5m2 0h1m1 0h1m2 0h1m5 0h1M1 13.5h2m1 0h1m1 0h2m1 0h1m4 0h2m3 0h3m3 0h1m2 0h1M3 14.5h1m5 0h1m1 0h3m3 0h2m1 0h1m1 0h2m2 0h2m3 0h2M1 15.5h1m1 0h8m1 0h1m1 0h1m2 0h1m1 0h2m1 0h1m3 0h1m2 0h1m2 0h1M3 16.5h2m1 0h1m2 0h1m3 0h1m6 0h4m1 0h1m2 0h3m1 0h2M1 17.5h3m1 0h3m1 0h1m1 0h1m4 0h1m4 0h7m1 0h5M1 18.5h1m1 0h4m4 0h1m1 0h1m3 0h4m1 0h1m1 0h1m1 0h1m1 0h1m2 0h1m1 0h1M3 19.5h2m2 0h2m1 0h1m2 0h2m2 0h1m1 0h2m3 0h1m3 0h2M1 20.5h1m3 0h2m1 0h1m1 0h1m2 0h2m5 0h2m1 0h4m1 0h1m3 0h1M1 21.5h7m1 0h1m2 0h6m5 0h1m2 0h2m2 0h1m1 0h2M1 22.5h2m1 0h1m5 0h1m2 0h3m3 0h4m2 0h1m1 0h4M3 23.5h1m2 0h3m1 0h2m4 0h1m1 0h1m2 0h8m1 0h3M3 24.5h1m1 0h1m4 0h1m3 0h1m1 0h4m2 0h1m1 0h1m1 0h1m4 0h1m1 0h1M1 25.5h2m2 0h4m3 0h2m1 0h1m1 0h1m1 0h5m1 0h5m1 0h3M9 26.5h1m2 0h2m6 0h2m1 0h3m3 0h1m1 0h3M1 27.5h7m1 0h3m4 0h1m1 0h1m2 0h3m1 0h1m1 0h1m1 0h2m1 0h2M1 28.5h1m5 0h1m3 0h2m1 0h6m2 0h1m2 0h1m3 0h1m3 0h1M1 29.5h1m1 0h3m1 0h1m1 0h1m2 0h1m6 0h1m1 0h1m3 0h6M1 30.5h1m1 0h3m1 0h1m4 0h1m1 0h1m2 0h2m1 0h1m2 0h2m2 0h1m1 0h1m1 0h1m1 0h1M1 31.5h1m1 0h3m1 0h1m2 0h1m1 0h2m3 0h1m1 0h2m1 0h1m1 0h1m1 0h2m1 0h1M1 32.5h1m5 0h1m2 0h2m8 0h5m1 0h2m2 0h1M1 33.5h7m1 0h1m3 0h1m2 0h1m5 0h3m1 0h1m3 0h2m1 0h1"/></svg>"`;
